### PR TITLE
only return head: null if source branch was deleted

### LINKED ISSUE #6613: [API] Pull Request is null when branch was deleted
- Gitea version (or commit ref): 3b612ce
- Git version: 2.20.1
- Operating system: Linux
- Database:
  - [x] PostgreSQL
  - [ ] MySQL
  - [ ] MSSQL
  - [ ] SQLite
- Can you reproduce the bug at https://try.gitea.io:
  - [x] Yes (provide example URL)
  - [ ] No
  - [ ] Not relevant

When a Pull Request was merged, the API endpoint `GET /api/v1/repo/<org>/<user>/pulls/<number>` returns null as soon as the branch is deleted (see [1]), while the web interface still lists all information about the PR (see [2]). Recreating the branch fixes the issue. The list of all pull-requests shows the same error (see [3]).

This was already reported in #4595, but the issue was closed because there was no activity.

Examples on `try.gitea.io`:
- [[1] faulty pull with missing branch](https://try.gitea.io/api/v1/repos/denny.crane/api-pulls-bug/pulls/1)
- [[2] faulty pull in web interface](https://try.gitea.io/denny.crane/api-pulls-bug/pulls/1)
- [[3] list of pulls](https://try.gitea.io/api/v1/repos/denny.crane/api-pulls-bug/pulls)

### PR DESCRIPTION
**Pull request by quantonganh**
Fix #6613

### CODE REVIEW

**Comment by lunny** [comment]
Could you add a test for it?

File: models/issue.go [hunk 1]
```diff
@@ -308,8 +308,8 @@ func (issue *Issue) APIURL() string {
 	return issue.Repo.APIURL() + "/" + path.Join("issues", fmt.Sprint(issue.Index))
 }
 
-// HTMLURL returns the absolute URL to this issue.
-func (issue *Issue) HTMLURL() string {
+// URL returns the absolute URL to this issue.
+func (issue *Issue) URL() string {
```
Comment (lafriks) [inline models/issue.go]: Why split to two functions if they return same data?

**Comment by quantonganh** [comment]
> Could you add a test for it?

https://github.com/go-gitea/gitea/blob/a2f014f65da9b6a75dd644bd9c6f9be812294ce3/models/pull_test.go#L34

I understand that to add this test, we have to:
- create a PR
- merge that PR
- delete source branch
- get PR and assert that it is not nil

```
func TestPullRequest_APIFormat(t *testing.T) {
    assert.NoError(t, PrepareTestDatabase())
    pr := AssertExistsAndLoadBean(t, &PullRequest{ID: 1}).(*PullRequest)
    var (
        ctx  *context.context
        form auth.MergePullRequestForm
    )
    assert.NoError(t, pr.Merge(ctx.User, ctx.Repo.GitRepo, MergeStyle(form.Do), pr.GetDefaultMergeMessage()))

    gitRepo, err := git.OpenRepository(pr.HeadRepo.RepoPath())
    assert.NoError(t, err)
    assert.NoError(t, gitRepo.DeleteBranch(pr.HeadBranch, git.DeleteBranchOptions{
        Force: true,
    }))
    assert.Nil(t, pr.HeadRepo)
    assert.NotNil(t, pr.APIFormat())
}
```

I'm stuck at the 2nd step: I don't know how to pass the params to Merge(): https://github.com/go-gitea/gitea/blob/a2f014f65da9b6a75dd644bd9c6f9be812294ce3/models/pull.go#L400

**Comment by quantonganh** [comment]
Build is passed: https://drone.gitea.io/go-gitea/gitea/9634

@lunny Could you please take a review?

**Review by lunny — CHANGES_REQUESTED** [review]

File: models/pull.go [hunk 2]
```diff
@@ -236,13 +211,48 @@ func (pr *PullRequest) apiFormat(e Engine) *api.PullRequest {
...
 		HasMerged: pr.HasMerged,
-		Base:      apiBaseBranchInfo,
-		Head:      apiHeadBranchInfo,
 		MergeBase: pr.MergeBase,
 		Deadline:  apiIssue.Deadline,
 		Created:   pr.Issue.CreatedUnix.AsTimePtr(),
 		Updated:   pr.Issue.UpdatedUnix.AsTimePtr(),
 	}
+	baseBranch, err = pr.BaseRepo.GetBranch(pr.BaseBranch)
+	if err != nil {
```
[the remark this answers is not in the document]
Comment (quantonganh) [reply models/pull.go:221]: Fixed: https://github.com/go-gitea/gitea/pull/6705/commits/cf3f71408011f633316157336785118fdd97f8bf

[the remark this answers is not in the document]
Comment (quantonganh) [reply models/pull.go:221]: How can I restart this (https://drone.gitea.io/go-gitea/gitea/9659/1/10) without pushing a new commit?

**Review by lunny — APPROVED** [review]

**Comment by lunny** [comment]
conflicted

**Comment by quantonganh** [comment]
> conflicted

@lunny rebased.

**Review by lafriks — APPROVED** [review]

**Comment by lafriks** [comment]
Should we backport this to 1.9?

**Comment by lunny** [comment]
I think we should.

**Comment by lafriks** [comment]
@quantonganh please send backport for commit ff85a63 to `release/v1.9` branch

**Comment by quantonganh** [comment]
@lafriks https://github.com/go-gitea/gitea/pull/7376

**Pull request merged**